File: sdk/typescript/src/exceptions.ts
/** Base error for all DNS of Money SDK errors. */
export class DNSOfMoneyError extends Error {
  readonly statusCode?: number;

  constructor(message: string, statusCode?: number) {
    super(message);
    this.name = "DNSOfMoneyError";
    this.statusCode = statusCode;
  }
}

/** Raised when a pay: alias does not exist (HTTP 404). */
export class AliasNotFoundError extends DNSOfMoneyError {
  constructor(alias: string) {
    super(`Alias not found: ${alias}`, 404);
    this.name = "AliasNotFoundError";
  }
}

/** Raised when a pay: alias is already registered (HTTP 409). */
export class AliasTakenError extends DNSOfMoneyError {
  constructor(alias: string) {
    super(`Alias already taken: ${alias}`, 409);
    this.name = "AliasTakenError";
  }
}

/** Raised when the API key is missing or invalid (HTTP 401). */
export class AuthenticationError extends DNSOfMoneyError {
  constructor() {
    super("Invalid or missing API key", 401);
    this.name = "AuthenticationError";
  }
}

/** Raised when the API rate limit is exceeded (HTTP 429). */
export class RateLimitError extends DNSOfMoneyError {
  readonly retryAfter?: number;

  constructor(retryAfter?: number) {
    const msg = retryAfter
      ? `Rate limit exceeded — retry after ${retryAfter}s`
      : "Rate limit exceeded";
    super(msg, 429);
    this.name = "RateLimitError";
    this.retryAfter = retryAfter;
  }
}

/** Raised when the founding tier cap (500) has been reached. */
export class CapReachedError extends DNSOfMoneyError {
  constructor() {
    super("Founding tier cap reached (500/500)", 409);
    this.name = "CapReachedError";
  }
}

File: sdk/python/dnsofmoney/exceptions.py
"""DNS of Money SDK — exception hierarchy."""


class DNSOfMoneyError(Exception):
    """Base exception for all DNS of Money SDK errors."""

    def __init__(self, message: str, status_code: int | None = None):
        self.status_code = status_code
        super().__init__(message)


class AliasNotFoundError(DNSOfMoneyError):
    """Raised when a pay: alias does not exist (HTTP 404)."""

    def __init__(self, alias: str):
        super().__init__(f"Alias not found: {alias}", status_code=404)


class AliasTakenError(DNSOfMoneyError):
    """Raised when a pay: alias is already registered (HTTP 409)."""

    def __init__(self, alias: str):
        super().__init__(f"Alias already taken: {alias}", status_code=409)


class AuthenticationError(DNSOfMoneyError):
    """Raised when the API key is missing or invalid (HTTP 401)."""

    def __init__(self):
        super().__init__("Invalid or missing API key", status_code=401)


class RateLimitError(DNSOfMoneyError):
    """Raised when the API rate limit is exceeded (HTTP 429)."""

    def __init__(self, retry_after: int | None = None):
        self.retry_after = retry_after
        msg = "Rate limit exceeded"
        if retry_after:
            msg += f" — retry after {retry_after}s"
        super().__init__(msg, status_code=429)


class CapReachedError(DNSOfMoneyError):
    """Raised when the founding tier cap (500) has been reached."""

    def __init__(self):
        super().__init__("Founding tier cap reached (500/500)", status_code=409)
